src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,281 @@
+                            _registeredWaitHandles[j] = null;
+                            _waitHandles[j + 1] = null;
+                            break;
+                        }
+                    }
+                    _pendingRemoves[i] = null;
+                }
+                _numPendingRemoves = 0;
+
+                // Fill in nulls
+                // This is O(1), Goes through each of the 63 possible handles once.
+                for (int i = 0; i < _numUserWaits; i++)
+                {
+                    if (_registeredWaitHandles[i] == null)
+                    {
+                        for (int j = _numUserWaits - 1; j > i; j--)
+                        {
+                            if(_registeredWaitHandles[j] != null)
+                            {
+                                _registeredWaitHandles[i] = _registeredWaitHandles[j];
+                                _registeredWaitHandles[j] = null;
+                                _numUserWaits = j;
+                                break;
+                            }
+                        }
+                        if (_registeredWaitHandles[i] == null)
+                        {
+                            _numUserWaits = i - 1;
+                            break;
+                        }
+                    }
+                }
+
+                // Recalculate Timeout
+                int timeout = Timeout.Infinite;
+                for (int i = 0; i < _numUserWaits; i++)
+                {
+                    if (timeout == Timeout.Infinite)
+                    {
+                        timeout = _registeredWaitHandles[i].Timeout;
+                    }
+                    else
+                    {
+                        timeout = Math.Min(_registeredWaitHandles[i].Timeout, timeout);
+                    }
+                }
+                _currentTimeout = timeout;
+                _registeredHandlesLock.Release();
+                _removesLock.Release();
+            }
+
+            private void QueueWaitCompletion(RegisteredWaitHandle registeredHandle, bool timedOut)
+            {
+                ThreadPool.QueueUserWorkItem(CompleteWait, new CompletedWaitHandle(registeredHandle, timedOut));
+            }
+
+            private void CompleteWait(object state)
+            {
+                CompletedWaitHandle handle = (CompletedWaitHandle)state;
+                handle.CompletedHandle.CanUnregister.Reset();
+                handle.CompletedHandle.PerformCallback(handle.TimedOut);
+                handle.CompletedHandle.CanUnregister.Set();
+                if (!handle.CompletedHandle.Repeating)
+                {
+                    QueueOrExecuteUnregisterWait(handle.CompletedHandle);
+                }
+            }
+
+            public bool RegisterWaitHandle(RegisteredWaitHandle handle)
+            {
+                StartWaitThreadIfNotStarted();
+
+                if (_numUserWaits == WaitHandle.MaxWaitHandles - 1)
```

Comment:
Is it really safe to check the value of _numUserWaits without of lock protection? This value can change by the time the code acquires the _registeredHandlesLock lock in AddWaitHandleForNextWait and you'll get an exception.